## Importing Data

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
# creating path for data import

path = r'C:\Users\scott\Desktop\Data Analysis\CF Immersive\Python\Jan-2022 Instacart Basket Analysis'

In [3]:
# importing products data

df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'products.csv'), index_col = False)

In [79]:
# importing orders data

df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

## Checking Products Data

### *Step 2 of Task

In [4]:
# checking shape of df_prods

df_prods.shape

(49693, 5)

In [80]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### Max value of prices column being 99999 is a possible concern

## Removing outliers in 'prices' column

In [7]:
df_prods.loc[df_prods['prices'] > 100, 'prices'] = np.nan

### Fixing mixed-type data

In [9]:
# creating data frame for fixing mixed-type columns

df_test = pd.DataFrame()
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
# checking for mixed-type data in test

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [9]:
# checking for mixed-type data in df_prods

for col in df_prods.columns.tolist():
    weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_prods[weird]) > 0:
        print (col)

product_name


In [10]:
# fixing data type in df_prods

df_prods['product_name'] = df_prods['product_name'].astype('str')

### Missing Values

In [81]:
# finding missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# view missing values

df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

Empty DataFrame
Columns: [product_id, product_name, aisle_id, department_id, prices]
Index: []

In [14]:
# creating clean products dataframe by filtering out missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape

(49693, 5)

### Duplicates

In [15]:
# finding full duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [16]:
# addressing duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49688, 5)

### Exporting Products Data

In [17]:
# exporting checked products data

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

## Checking Orders Data

### Checking describe function

In [87]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

### Finding mixed-type data

In [88]:
# checking for mixed-type data

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)
        
# there is no mixed-type columns

### Missing Values

In [89]:
# checking for missing values

df_ords.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

#### 'days_since_last_order' has 206209 missing values out of 3421083 (~ 6%).  Because these missing values correspond to customers with only one order, 'days_since_last_order' cannot be calculated. These missing values will be left as is.

In [93]:
# filtering df_ords by order_number = 1 to show that these are missing 'days_since_last_order'

df_ords_order_1 = df_ords['order_number'] == 1
df_ords[df_ords_order_1].describe()

Unnamed: 0      order_id        user_id  order_number  \
count  2.062090e+05  2.062090e+05  206209.000000      206209.0   
mean   1.712637e+06  1.708462e+06  103105.000000           1.0   
std    9.874816e+05  9.881299e+05   59527.555167           0.0   
min    0.000000e+00  2.000000e+01       1.000000           1.0   
25%    8.579730e+05  8.507300e+05   51553.000000           1.0   
50%    1.717068e+06  1.706246e+06  103105.000000           1.0   
75%    2.570518e+06  2.564292e+06  154657.000000           1.0   
max    3.421069e+06  3.421081e+06  206209.000000           1.0   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count       206209.000000      206209.000000                    0.0  
mean             2.754118          13.626597                    NaN  
std              2.076205           4.223769                    NaN  
min              0.000000           0.000000                    NaN  
25%              1.000000          11.000000                    NaN  
50%              3.000000          14.000000                    NaN  
75%              5.000000          17.000000                    NaN  
max              6.000000          23.000000                    NaN

### Duplicates

In [104]:
# finding full duplicates

df_ords_dups = df_ords[df_ords.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

#### There are no duplicates in the orders data

### Exporting orders data

In [97]:
# exporting checked orders data

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))